### DISCUSSION TITLE
wip: new `CCParse`

### DISCUSSION THREAD

**Discussion by c-cube** [c-cube/ocaml-containers#381] [General]
I'm working, in #380, on a significant refactoring of `CCParse`, to try and make it easier to parse basic things without writing a .mll and .mly as menhir would require. That can be very basic line oriented formats, IRC logs, competitive programming input problems, etc.

I'd love to get some feedback on the new functions and the changes in how backtracking works. @anuragsoni I think we discussed this recently, in case you're still interested :)

edit: some examples:
- [a very simple irc log parser](https://github.com/c-cube/ocaml-containers/blob/wip-parse-2021-05-04/examples/ccparse_irclogs.ml) illustrating slices and `recurse`
- [a basic S-expression parser](https://github.com/c-cube/ocaml-containers/blob/wip-parse-2021-05-04/examples/ccparse_sexp.ml) which illustrates recursion and handles quoted strings with escape sequences

**Comment by Gbury** [comment]
I haven't yet had enough time to look at it in detail, but from what I can tell, there doesn't seem to be an easy way to attach locations to the result of the parsing functions, i.e. locations are only available in the error cases, and not in the success ones.

**Reply by c-cube** [reply]
Well, this is not designed to write Dolmen, with locations on each node of a sophisticated AST. It's designed for basic parsers, where I don't think locations are important?

**Reply by Gbury** [reply]
That's probably reasonable, I just wanted to make sure that was a conscious decision, ^^

**Comment by anuragsoni** [comment]
#380 looks pretty nice! I'll take a closer look at the changes soon, but at a glance I like the proposed api!

**Reply by c-cube** [reply]
Curious to hear feedback on the `slice` thing, which is quite new :)
I think it's relatively unusual compared to other parser combinator libraries, afaict.
